### DISCUSSION TITLE
Why have separate `validate_op` and `apply`?

### DISCUSSION THREAD

**Discussion by Kiiyya** [spacejam/rust-crdt#125] [Q&A]
Wouldn't it make more sense for `apply()` to return a `Result<(), Validation>` instead, and not have any `validate_op` at all?

**Comment by davidrusu** [comment]
There are instances in real world applications where we want to validate an Op without applying it. For example, in the byzantine-reliable-broadcast protocol we separate out the validation phase from the applying phase: https://github.com/maidsafe/brb_dt_orswot/blob/master/src/brb_orswot.rs#L99-L131

2-phase commit follows a similar pattern of PREPARE followed by COMMIT. It's a common pattern across distributed algorithms so we want to be flexible in our API's here to support these cases rather than monolithic functions that do it all.

**Reply by Kiiyya** [reply]
Ah, looks like I have to do some reading on byzantine reliable broadcast then. Thanks for the answer!
I'll accept your answer once GitHub fixes its [bug](https://github.com/github/feedback/discussions/2903#discussioncomment-1220341)...
